hack/helm/release/internal/helm/google_storage_test.go

```diff
@@ -0,0 +1,313 @@
+		"metageneration": "1",
+		"contentType": "text/plain; charset=utf-8",
+		"storageClass": "STANDARD",
+		"size": "67750",
+		"md5Hash": "0zxTfawe/vFq8rZX9Aa4bg==",
+		"crc32c": "i1Pc8w==",
+		"etag": "CN+Mjofdj/4CEAE=",
+		"timeCreated": "2023-04-04T07:42:22.813Z",
+		"updated": "2023-04-04T07:42:22.813Z",
+		"timeStorageClassUpdated": "2023-04-04T07:42:22.813Z"
+	  }`
+	// valid error response format for google storage JSON API.
+	errorResponseFmt = `{
+		"error": {
+		 "errors": [
+		  {
+		   "domain": "global",
+		   "reason": "conditionNotMet",
+		   "message": "At least one of the pre-conditions you specified did not hold.",
+		   "locationType": "header",
+		   "location": "If-Match"
+		  }
+		 ],
+		 "code": %d,
+		 "message": "At least one of the pre-conditions you specified did not hold."
+		 }
+		}`
+)
+
+func Test_readIndexFromBucket(t *testing.T) {
+	tmpdir, err := os.MkdirTemp(os.TempDir(), "helm_google_storage_test")
+	if err != nil {
+		t.Fatalf("while creating temp dir: %s", err)
+		return
+	}
+	defer func() {
+		os.RemoveAll(tmpdir)
+	}()
+	type args struct {
+		ctx    context.Context
+		config readIndexConfig
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    *index
+		wantErr bool
+	}{
+		{
+			name: "happy path",
+			args: args{
+				ctx: context.Background(),
+				config: readIndexConfig{
+					indexFile: filepath.Join(tmpdir, oldIndexFileName),
+					bucket:    testBucketName,
+				},
+			},
+			want: &index{
+				generation: validGeneration,
+				path:       filepath.Join(tmpdir, oldIndexFileName),
+			},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			client, ts := newTestStorageClient(t)
+			if ts != nil {
+				defer ts.Close()
+			}
+			tt.args.config.client = client
+			got, err := readIndexFromBucket(tt.args.ctx, tt.args.config)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("readIndexFromBucket() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("readIndexFromBucket() = %v, want %v", got, tt.want)
+			}
+			if tt.want != nil {
+				if _, err := os.Stat(tt.want.path); os.IsNotExist(err) {
+					t.Errorf("readIndexFromBucket() = expecting file at %s but none was found", tt.want.path)
+				}
+			}
+		})
+	}
+}
+
+func newTestStorageClient(t *testing.T) (*storage.Client, *httptest.Server) {
+	t.Helper()
+	ts := testServer()
+	client, err := storage.NewClient(
+		context.Background(),
+		option.WithEndpoint(ts.URL),
+		option.WithCredentials(&google.Credentials{
+			ProjectID: "fake",
+			TokenSource: oauth2.StaticTokenSource(&oauth2.Token{
+				AccessToken: "fake",
+				Expiry:      time.Now().Add(1 * time.Hour),
+			}),
+		}))
+	if err != nil {
+		t.Fatalf("getting new storage client: %s", err)
+		return nil, nil
+	}
+	return client, ts
+}
+
+func testServer() *httptest.Server {
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
The below set of if statements aren't very readable. Any suggestions on cleaning this up would be appreciated.